MCXN94X Adding LPTMR Counter Support

EmilioCBen opened the pull request.

EmilioCBen committed:
dts: arm: nxp_mcxn947: Added lptmr Node in dts

Added a single instance lptmr node on the
mcxn947 soc dts. Updated counter lptmr to
have max value property.

Signed-off-by: Emilio Benavente <emilio.benavente@nxp.com>

EmilioCBen committed:
boards: nxp: frdm_mcxn947: Enable lpmtr for frdm_mcxn947

Enabled one instance of lptmr
running from the 16k FRO.

Signed-off-by: Emilio Benavente <emilio.benavente@nxp.com>

EmilioCBen committed:
tests: drivers: counter: counter_basic_api: enabled lptmr testing

Enabled the counter_basic_api test to test lptmr.

Signed-off-by: Emilio Benavente <emilio.benavente@nxp.com>

butok commented on dts/arm/nxp/nxp_mcxn94x_common.dtsi:
```diff
@@ -850,6 +850,15 @@
 	};
+
+	lptmr0: lptmr@4a000 {
+		compatible = "nxp,kinetis-lptmr";
```
Can it be "nxp,lptmr" without kinetis?

danieldegrasse replied:
Unfortunately no, not unless we change the compatible string across all parts. Typically we have left compatibles with the family name as is to avoid churn for downstream driver users, although we want to avoid adding new drivers with the family name (so this issue does not occur in the future)

butok replied:
hope, we will be brave and clean it up sooner.

danieldegrasse commented on boards/nxp/frdm_mcxn947/frdm_mcxn947_mcxn947_cpu0.yaml:
```diff
@@ -27,4 +27,5 @@ supported:
   - regulator
   - adc
   - usb_device
+  - lpmtr
```
This yaml definition won't trigger any CI tests, as none have dependency on an IP name like `lptmr`. The tag you need to get counter tests to run is `counter`, which is already present

danieldegrasse commented on dts/arm/nxp/nxp_mcxn94x_common.dtsi:
Commit adding the LPTMR to MCX DTS should come after the commits renaming the LPTMR binding, and should use the new `nxp,lptmr` binding name

danieldegrasse commented on boards/nxp/frdm_mcxn947/board.c:
> boards: nxp: frdm_mcxn947

> Enabled one instance of lptmr
> running from the 16k FRO.

Commit title here needs to be more descriptive

danieldegrasse commented on drivers/timer/mcux_lptmr_timer.c:
This change should be in a separate commit from the updates to the counter driver

EmilioCBen committed:
drivers: counter: counter_mcux_lptmr: Update to start with top value

Updated the driver to start with a value to compare the counter
with otherwise the counter will not start until the user sets
the top value manually, an issue that will occur inside the counter
test.

Signed-off-by: Emilio Benavente <emilio.benavente@nxp.com>

decsny commented on drivers/timer/Kconfig.mcux_lptmr:
```diff
@@ -6,9 +6,9 @@
 	default y
-	depends on DT_HAS_NXP_KINETIS_LPTMR_ENABLED
+	depends on DT_HAS_NXP_LPTMR_ENABLED
+	depends on !DT_HAS_NXP_OS_TIMER_ENABLED
```
this seems SOC specific maybe it should go in the SOC defconfig

decsny commented on drivers/counter/counter_mcux_lptmr.c:
```diff
@@ -210,7 +212,11 @@ static void mcux_lptmr_irq_config_0(const struct device *dev);
 static struct mcux_lptmr_config mcux_lptmr_config_0 = {
 	.info = {
+#if defined(CONFIG_SOC_SERIES_MCXNX4X)
+		.max_top_value = UINT32_MAX,
```
Looks like this needs to come from DT

EmilioCBen committed:
drivers: timer: updated lptmr_timer binding

Updated the lptmr timer binding from
nxp,kinetis-lptmr to nxp,lptmr.

Signed-off-by: Emilio Benavente <emilio.benavente@nxp.com>

decsny commented on dts/bindings/counter/nxp,lptmr.yaml:
```diff
@@ -29,3 +29,8 @@ properties:
     type: boolean
     description: Pulse counter input pin is active-low
+
+  max-top-value:
```
might make more sense to put the resolution of the counter in bits

decsny commented on dts/bindings/counter/nxp,lptmr.yaml:
```diff
@@ -29,3 +29,8 @@ properties:
     type: boolean
     description: Pulse counter input pin is active-low
+
+  max-resolution:
```
"max-resolution" does not make sense, just call it "resolution"

decsny commented on dts/bindings/counter/nxp,lptmr.yaml:
```diff
@@ -29,3 +29,8 @@ properties:
+  max-resolution:
+    type: int
+    required: true
+    description: Represents the maximum width of the timer in bits.
```
no maximum

decsny commented:
reassigned to david because there is not really much of anything new here it's just an enablement of something already in tree on an nxp platform
